File: go/indicators/perrykaufman/kaufmanadaptivemovingaverage/kaufmanadaptivemovingaverage.go
package kaufmanadaptivemovingaverage

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// KaufmanAdaptiveMovingAverage (Kaufman's adaptive moving average, KAMA) is an EMA with the smoothing
// factor, α, being changed with each new sample within the fastest and the slowest boundaries:
//
// KAMAᵢ = αPᵢ + (1 - α)*KAMAᵢ₋₁,  α = (αs + (αf - αs)ε)²
//
// where the αf is the α of the fastest (shortest, default 2 samples) period boundary,
// the αs is the α of the slowest (longest, default 30 samples) period boundary,
// and ε is the efficiency ratio:
//
// ε = |P - Pℓ| / ∑|Pᵢ - Pᵢ₊₁|,  i ≤ ℓ-1
//
// where ℓ is a number of samples used to calculate the ε.
// The recommended values of ℓ are in the range of 8 to 10.
//
// The efficiency ratio has the value of 1 when samples move in the same direction for
// the full ℓ periods, and a value of 0 when samples are unchanged over the ℓ periods.
// When samples move in wide swings within the interval, the sum of the denominator
// becomes very large compared with the numerator and the ε approaches 0.
// Smaller values of ε result in a smaller smoothing constant and a slower trend.
//
// The indicator is not primed during the first ℓ updates.
//
// Reference:
// Perry J. Kaufman, Smarter Trading, McGraw-Hill, Ney York, 1995, pp. 129-153.
type KaufmanAdaptiveMovingAverage struct {
	mu sync.RWMutex
	core.LineIndicator
	efficiencyRatioLength int
	windowCount           int
	window                []float64
	absoluteDelta         []float64
	absoluteDeltaSum      float64
	alphaFastest          float64
	alphaSlowest          float64
	alphaDiff             float64
	value                 float64
	efficiencyRatio       float64
	primed                bool
}

// NewKaufmanAdaptiveMovingAverageLength returns an instnce of the indicator
// created using supplied parameters based on length.
func NewKaufmanAdaptiveMovingAverageLength(
	p *KaufmanAdaptiveMovingAverageLengthParams,
) (*KaufmanAdaptiveMovingAverage, error) {
	return newKaufmanAdaptiveMovingAverage(p.EfficiencyRatioLength,
		p.FastestLength, p.SlowestLength,
		math.NaN(), math.NaN(),
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

// NewKaufmanAdaptiveMovingAverageSmoothingFactor returns an instnce of the indicator
// created using supplied parameters based on smoothing factor.
func NewKaufmanAdaptiveMovingAverageSmoothingFactor(
	p *KaufmanAdaptiveMovingAverageSmoothingFactorParams,
) (*KaufmanAdaptiveMovingAverage, error) {
	return newKaufmanAdaptiveMovingAverage(p.EfficiencyRatioLength,
		0, 0,
		p.FastestSmoothingFactor, p.SlowestSmoothingFactor,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

//nolint:funlen,cyclop
func newKaufmanAdaptiveMovingAverage(efficiencyRatioLength int,
	fastestSmoothingLength int, slowestSmoothingLength int,
	fastestSmoothingFactor float64, slowestSmoothingFactor float64,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*KaufmanAdaptiveMovingAverage, error) {
	const (
		invalid = "invalid Kaufman adaptive moving average parameters"
		fmtl    = "%s: %s length should be larger than 1"
		fmta    = "%s: %s smoothing factor should be in range [0, 1]"
		fmtw    = "%s: %w"
		fmtnl   = "kama(%d, %d, %d%s)"
		fmtna   = "kama(%d, %.4f, %.4f%s)"
		two     = 2
		epsilon = 0.00000001
	)

	var (
		mnemonic  string
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if efficiencyRatioLength < two {
		return nil, fmt.Errorf(fmtl, invalid, "efficiency ratio")
	}

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if math.IsNaN(fastestSmoothingFactor) { //nolint:nestif
		if fastestSmoothingLength < two {
			return nil, fmt.Errorf(fmtl, invalid, "fastest smoothing")
		}

		if slowestSmoothingLength < two {
			return nil, fmt.Errorf(fmtl, invalid, "slowest smoothing")
		}

		fastestSmoothingFactor = float64(two) / float64(1+fastestSmoothingLength)
		slowestSmoothingFactor = float64(two) / float64(1+slowestSmoothingLength)

		mnemonic = fmt.Sprintf(fmtnl, efficiencyRatioLength,
			fastestSmoothingLength, slowestSmoothingLength,
			core.ComponentTripleMnemonic(bc, qc, tc))
	} else {
		if fastestSmoothingFactor < 0. || fastestSmoothingFactor > 1. {
			return nil, fmt.Errorf(fmta, invalid, "fastest")
		}

		if slowestSmoothingFactor < 0. || slowestSmoothingFactor > 1. {
			return nil, fmt.Errorf(fmta, invalid, "slowest")
		}

		if fastestSmoothingFactor < epsilon {
			fastestSmoothingFactor = epsilon
		}

		if slowestSmoothingFactor < epsilon {
			slowestSmoothingFactor = epsilon
		}

		mnemonic = fmt.Sprintf(fmtna, efficiencyRatioLength,
			fastestSmoothingFactor, slowestSmoothingFactor,
			core.ComponentTripleMnemonic(bc, qc, tc))
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	desc := "Kaufman adaptive moving average " + mnemonic

	kama := &KaufmanAdaptiveMovingAverage{
		efficiencyRatioLength: efficiencyRatioLength,
		alphaFastest:          fastestSmoothingFactor,
		alphaSlowest:          slowestSmoothingFactor,
		alphaDiff:             fastestSmoothingFactor - slowestSmoothingFactor,
		value:                 math.NaN(),
		efficiencyRatio:       math.NaN(),

		// These slices will be automatically filled with zeroes.
		window:        make([]float64, efficiencyRatioLength+1),
		absoluteDelta: make([]float64, efficiencyRatioLength+1),
	}

	kama.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, kama.Update)

	return kama, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *KaufmanAdaptiveMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
// It always has a single scalar output -- the calculated value of the moving average.
func (s *KaufmanAdaptiveMovingAverage) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.KaufmanAdaptiveMovingAverage,
		Mnemonic:    s.LineIndicator.Mnemonic,
		Description: s.LineIndicator.Description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(KaufmanAdaptiveMovingAverageValue),
				Type:        outputs.ScalarType,
				Mnemonic:    s.LineIndicator.Mnemonic,
				Description: s.LineIndicator.Description,
			},
		},
	}
}

// Update updates the value of the moving average given the next sample.
func (s *KaufmanAdaptiveMovingAverage) Update(sample float64) float64 { //nolint:funlen
	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	const epsilon = 0.00000001

	var temp float64

	if s.primed { //nolint:nestif
		temp = math.Abs(sample - s.window[s.efficiencyRatioLength])
		s.absoluteDeltaSum += temp - s.absoluteDelta[1]

		for i := range s.efficiencyRatioLength {
			j := i + 1
			s.window[i] = s.window[j]
			s.absoluteDelta[i] = s.absoluteDelta[j]
		}

		s.window[s.efficiencyRatioLength] = sample
		s.absoluteDelta[s.efficiencyRatioLength] = temp
		delta := math.Abs(sample - s.window[0])

		if s.absoluteDeltaSum <= delta || s.absoluteDeltaSum < epsilon {
			temp = 1.0
		} else {
			temp = delta / s.absoluteDeltaSum
		}

		s.efficiencyRatio = temp
		temp = s.alphaSlowest + temp*s.alphaDiff
		s.value += (sample - s.value) * temp * temp

		return s.value
	} else {
		s.window[s.windowCount] = sample
		if 0 < s.windowCount {
			temp = math.Abs(sample - s.window[s.windowCount-1])
			s.absoluteDelta[s.windowCount] = temp
			s.absoluteDeltaSum += temp
		}

		if s.efficiencyRatioLength == s.windowCount {
			s.primed = true
			delta := math.Abs(sample - s.window[0])

			if s.absoluteDeltaSum <= delta || s.absoluteDeltaSum < epsilon {
				temp = 1.0
			} else {
				temp = delta / s.absoluteDeltaSum
			}

			s.efficiencyRatio = temp
			temp = s.alphaSlowest + temp*s.alphaDiff
			s.value = s.window[s.efficiencyRatioLength-1]
			s.value += (sample - s.value) * temp * temp

			return s.value
		} else {
			s.windowCount++
		}
	}

	return math.NaN()
}

File: ts/indicators/perry-kaufman/kaufman-adaptive-moving-average/kaufman-adaptive-moving-average.ts
import { buildMetadata } from '../../core/build-metadata';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { KaufmanAdaptiveMovingAverageLengthParams } from './params';
import { KaufmanAdaptiveMovingAverageSmoothingFactorParams } from './params';

const guardLength = (object: any): object is KaufmanAdaptiveMovingAverageLengthParams => 'fastestLength' in object;

/** Function to calculate mnemonic of a __KaufmanAdaptiveMovingAverage__ indicator. */
export const kaufmanAdaptiveMovingAverageMnemonic =
  (params: KaufmanAdaptiveMovingAverageLengthParams | KaufmanAdaptiveMovingAverageSmoothingFactorParams): string => {
    if (guardLength(params)) {
      const p = params as KaufmanAdaptiveMovingAverageLengthParams;
      return 'kama('.concat(Math.floor(p.efficiencyRatioLength).toString(),
        ', ', Math.floor(p.fastestLength).toString(), ', ', Math.floor(p.slowestLength).toString(),
        componentTripleMnemonic(p.barComponent, p.quoteComponent, p.tradeComponent), ')');
    } else {
      const p = params as KaufmanAdaptiveMovingAverageSmoothingFactorParams;
      return 'kama('.concat(Math.floor(p.efficiencyRatioLength).toString(),
        ', ', p.fastestSmoothingFactor.toFixed(4), ', ', p.slowestSmoothingFactor.toFixed(4),
        componentTripleMnemonic(p.barComponent, p.quoteComponent, p.tradeComponent), ')');
    }
  };

/** __Kaufman Adaptive Moving Average__ (_KAMA_) is an EMA with the smoothing
 * factor, α, being changed with each new sample within the fastest and the slowest boundaries:
 *
 * KAMAᵢ = αPᵢ + (1 - α)*KAMAᵢ₋₁,  α = (αs + (αf - αs)ε)²
 *
 * where the αf is the α of the fastest (shortest, default 2 samples) period boundary,
 * the αs is the α of the slowest (longest, default 30 samples) period boundary,
 * and ε is the efficiency ratio:
 *
 * ε = |P - Pℓ| / ∑|Pᵢ - Pᵢ₊₁|,  i ≤ ℓ-1
 *
 * where ℓ is a number of samples used to calculate the ε.
 * The recommended values of ℓ are in the range of 8 to 10.
 *
 * The efficiency ratio has the value of 1 when samples move in the same direction for
 * the full ℓ periods, and a value of 0 when samples are unchanged over the ℓ periods.
 * When samples move in wide swings within the interval, the sum of the denominator
 * becomes very large compared with the numerator and the ε approaches 0.
 * Smaller values of ε result in a smaller smoothing constant and a slower trend.
 *
 * The indicator is not primed during the first ℓ updates.
 *
 * Reference:
 * Perry J. Kaufman, Smarter Trading, McGraw-Hill, Ney York, 1995, pp. 129-153.
 */
export class KaufmanAdaptiveMovingAverage extends LineIndicator {
  private readonly efficiencyRatioLength: number;
  private readonly alphaFastest: number;
  private readonly alphaSlowest: number;
  private readonly alphaDiff: number;
  private readonly window: Array<number>;
  private readonly absoluteDelta: Array<number>;
  private absoluteDeltaSum = 0;
  private value = 0;
  private efficiencyRatioValue = 0;
  private windowCount = 0;

  /**
   * Constructs an instance given a length in samples or a smoothing factor in (0, 1).
   **/
  public constructor(params: KaufmanAdaptiveMovingAverageLengthParams | KaufmanAdaptiveMovingAverageSmoothingFactorParams) {
    super();

    if (guardLength(params)) {
      const p = params as KaufmanAdaptiveMovingAverageLengthParams;

      this.efficiencyRatioLength = Math.floor(p.efficiencyRatioLength);
      if (this.efficiencyRatioLength < 2) {
        throw new Error('efficiency ratio length should be greater than 1');
      }

      const fastestLen = Math.floor(p.fastestLength);
      if (fastestLen < 2) {
        throw new Error('fastest length should be greater than 1');
      }

      const slowestLen = Math.floor(p.slowestLength);
      if (slowestLen < 2) {
        throw new Error('slowest length should be greater than 1');
      }

      this.alphaFastest = 2 / (fastestLen + 1);
      this.alphaSlowest = 2 / (slowestLen + 1);
    } else {
      const p = params as KaufmanAdaptiveMovingAverageSmoothingFactorParams;

      this.efficiencyRatioLength = Math.floor(p.efficiencyRatioLength);
      if (this.efficiencyRatioLength < 2) {
        throw new Error('efficiency ratio length should be greater than 1');
      }

      if (p.fastestSmoothingFactor <= 0 || p.fastestSmoothingFactor >= 1) {
        throw new Error('fastest smoothing factor should be in range (0, 1)');
      }

      if (p.slowestSmoothingFactor <= 0 || p.slowestSmoothingFactor >= 1) {
        throw new Error('slowest smoothing factor should be in range (0, 1)');
      }

      this.alphaFastest = p.fastestSmoothingFactor;
      this.alphaSlowest = p.slowestSmoothingFactor;
    }

    this.alphaDiff = this.alphaFastest - this.alphaSlowest;
    this.window = new Array<number>(this.efficiencyRatioLength + 1);
    this.absoluteDelta = new Array<number>(this.efficiencyRatioLength + 1);

    this.mnemonic = kaufmanAdaptiveMovingAverageMnemonic(params);
    this.description = 'Kaufman adaptive moving average ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;
    this.primed = false;
  }

  /** The value of the efficiency ratio component of the indicator. */
  public getEfficiencyRatio(): number { return this.primed ? this.efficiencyRatioValue : Number.NaN; }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.KaufmanAdaptiveMovingAverage,
      this.mnemonic,
      this.description,
      [
        { mnemonic: this.mnemonic, description: this.description },
      ],
    );
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    const epsilon = 0.00000001;

    let temp;

    if (this.primed) {
      temp = Math.abs(sample - this.window[this.efficiencyRatioLength]);
      this.absoluteDeltaSum += temp - this.absoluteDelta[1];

      for (let i = 0; i < this.efficiencyRatioLength; i++) {
        const j = i + 1;
        this.window[i] = this.window[j];
        this.absoluteDelta[i] = this.absoluteDelta[j];
      }

      this.window[this.efficiencyRatioLength] = sample;
      this.absoluteDelta[this.efficiencyRatioLength] = temp;
      const delta = Math.abs(sample - this.window[0]);

      if (this.absoluteDeltaSum <= delta || this.absoluteDeltaSum < epsilon) {
        temp = 1;
      } else {
        temp = delta / this.absoluteDeltaSum;
      }

      this.efficiencyRatioValue = temp;
      temp = this.alphaSlowest + temp * this.alphaDiff;
      this.value += (sample - this.value) * temp * temp;

      return this.value;
    } else { // Not primed.
      this.window[this.windowCount] = sample;

      if (0 < this.windowCount) {
        temp = Math.abs(sample - this.window[this.windowCount - 1]);
        this.absoluteDelta[this.windowCount] = temp;
        this.absoluteDeltaSum += temp;
      }

      if (this.efficiencyRatioLength === this.windowCount) {
        this.primed = true;
        const delta = Math.abs(sample - this.window[0]);

        if (this.absoluteDeltaSum <= delta || this.absoluteDeltaSum < epsilon) {
          temp = 1;
        } else {
          temp = delta / this.absoluteDeltaSum;
        }

        this.efficiencyRatioValue = temp;
        temp = this.alphaSlowest + temp * this.alphaDiff;
        this.value = this.window[this.efficiencyRatioLength - 1];
        this.value += (sample - this.value) * temp * temp;

        return this.value;
      } else {
        this.windowCount++;
      }
    }

    return Number.NaN;
  }
}
